### ISSUE TITLE
What is the official way of initializing Timber?

### ISSUE THREAD

**Issue by acobster** [timber/timber#2404]
## Is your feature request related to a problem? Please describe.

I found that now that Timber is a composer package only, I need to call `new Timber\Timber()` explicitly from my theme for anything to work. Is that the official way to initialize Timber now?

Also, do we want to refactor that process at all? It's a little awkward that the constructor calls `self::init()` statically, for example.

**Comment by gchtr** [comment]
We could probably improve that and not do anything in the constructor, because obviously that is a pattern that leads to other problems.

However, we can’t init Timber automatically, because of compatibility issues with environments like Bedrock. With Bedrock, Composer initializes earlier, when WordPress is not loaded yet, which would lead to fatal errors. More details in https://github.com/timber/timber/issues/979#issuecomment-220094476 and https://github.com/timber/timber/pull/1041.

How about `Timber::init()` as the official way to initialize Timber?

**Comment by twentyfortysix** [comment]
..how is it with initialisation of Timber with Bedrock now? I went thru all tickets I found and not even the tickets mentioned earlier helped me.
I have the Bedrock with Timber installed form github as library and following in functions.php in theme and I still get fatal error 
```
require_once( $_SERVER['DOCUMENT_ROOT'].'/../vendor/autoload.php' );
use Timber\Timber;

```

`Fatal error: Uncaught Error: Class 'Timber\Timber' not found...`

**Comment by jameelmoses** [comment]
> ..how is it with initialisation of Timber with Bedrock now? I went thru all tickets I found and not even the tickets mentioned earlier helped me.
> I have the Bedrock with Timber installed form github as library and following in functions.php in theme and I still get fatal error
> 
> ```
> require_once( $_SERVER['DOCUMENT_ROOT'].'/../vendor/autoload.php' );
> use Timber\Timber;
> ```
> 
> `Fatal error: Uncaught Error: Class 'Timber\Timber' not found...`

I'd need more info of your exact directory structure, but assuming you have Timber in your overall project's composer.json as opposed to a composer.json in the theme, this is what I have in my wp-config.php: `require_once dirname( __DIR__ ) . '/vendor/autoload.php';`

**Comment by twentyfortysix** [comment]
Hi jameelmoses

here is my composer.json
https://gist.github.com/twentyfortysix/5b88880c4e8e65ac7d012ff45a2b8294
Yes I try to load Timber as Library not as a plugin or use the composer inside the template directory.

As far as I remember we are discouraged to add anything to wp-config.php and use the application.php instead.. that is what the Bedrock documentation says.
It is not even needed as the wp-config.php made by Bedrock has those lines there already

```
require_once dirname(__DIR__) . '/vendor/autoload.php';
require_once dirname(__DIR__) . '/config/application.php';
require_once ABSPATH . 'wp-settings.php';
```

based on those lines I suppose(and I might be wrong indeed ) that the Timber as library is loaded already and we need nothing extra and use the Timber right away as it is loaded already... I believe.

**Comment by acobster** [comment]
@twentyfortysix

```
require_once( $_SERVER['DOCUMENT_ROOT'].'/../vendor/autoload.php' );
use Timber\Timber;
```

That code alone will not throw a fatal error, since `use` by itself does not actually load the class. Do you have a stack trace or a file/line number where that error is being triggered?

As of 2.x **alpha** the way to initialize it is `new Timber\Timber()` but that will change to `Timber::init()` before 2.0 lands. But this seems to be an issue with your composer setup. If it can't find the class, neither way will work.

**Comment by twentyfortysix** [comment]
here are first two lines of my functions.php

```
require_once( $_SERVER['DOCUMENT_ROOT'].'/../vendor/autoload.php' );
new Timber\Timber();
```
which throws this

`Fatal error: Uncaught Error: Class 'Timber\Timber' not found in /home/fprocesu/public_html/spunt.fprocesu.cz/bedrock/web/app/themes/almond-milk/functions.php:3 Stack trace: #0 /home/fprocesu/public_html/spunt.fprocesu.cz/bedrock/web/wp/wp-settings.php(546): include() #1 /home/fprocesu/public_html/spunt.fprocesu.cz/bedrock/web/wp-config.php(9): require_once('/home/fprocesu/...') #2 /home/fprocesu/public_html/spunt.fprocesu.cz/bedrock/web/wp/wp-load.php(55): require_once('/home/fprocesu/...') #3 /home/fprocesu/public_html/spunt.fprocesu.cz/bedrock/web/wp/wp-blog-header.php(13): require_once('/home/fprocesu/...') #4 /home/fprocesu/public_html/spunt.fprocesu.cz/bedrock/web/index.php(6): require('/home/fprocesu/...') #5 {main} thrown in /home/fprocesu/public_html/spunt.fprocesu.cz/bedrock/web/app/themes/almond-milk/functions.php on line 3`

maybe there is a mismatch in timber versions 
this is what I have in composer.json
```
{
      "type": "package",
      "package": {
        "name": "timber/timber",
        "version": "2.0.0",
        "type": "library",
        "source": {
          "url": "https://github.com/timber/timber.git",
          "type": "git",
          "reference": "master"
        }
      }
    }
```

**Comment by acobster** [comment]
The version installed when I test with your full composer.json (from the gist) is 1.18.2, so there does seem to be some kind of mismatch. However, that doesn't explain the class not found error, since the namespace and path of the `Timber\Timber` class haven't changed.

Is it possible you have multiple `vendor/autoload.php` files and you're requiring the wrong one? What is `$_SERVER['DOCUMENT_ROOT']`?

**Comment by twentyfortysix** [comment]
Hi
the $_SERVER['DOCUMENT_ROOT'] is /home/fprocesu/public_html/spunt.fprocesu.cz/bedrock/web
and so the autoload I'm loading is the /home/fprocesu/public_html/spunt.fprocesu.cz/bedrock/vendor/autoload.php

all in all there are these autoload files
./web/app/plugins/advanced-access-manager/vendor/autoload.php
./web/wp/wp-includes/sodium_compat/autoload.php
./vendor/squizlabs/php_codesniffer/autoload.php
./vendor/autoload.php

Just to be sure I have deleted the advanced-access-manager plugin. The problem remains.

**Comment by twentyfortysix** [comment]
hmm. this is weird.
When I run the composer require timber/timber within the theme directory (which is not what made it work.. see bellow)
it askes me if I want to use the main composer.json in the bedrock root, and I say yes and it starts to work.
Timber is installed in the bedrock/vendor directory. composer.json is updated to require/"timber/timber": "^1.18"

So I guess the ^2.0.0 version from github I defined in composer.json has some inconsistency. Or I defined it wrong.
Therefore this works for now

```
"require": {
    "php": ">=7.1",
    "composer/installers": "^1.11",
    "vlucas/phpdotenv": "^5.3",
    "oscarotero/env": "^2.1",
    "roots/bedrock-autoloader": "^1.0",
    "roots/bedrock-disallow-indexing": "^2.0",
    "roots/wordpress": "5.8",
    "roots/wp-config": "1.0.0",
    "roots/wp-password-bcrypt": "1.0.0",
    "timber/timber": "^1.18"
  },
```

**Comment by gchtr** [comment]
Resolved in #2549.
